src/libs/elektra/key.c

```diff
@@ -244,218 +244,220 @@ Key * keyVNew (const char * name, va_list va)
 
-	if (source->data.v)
+	// duplicate dynamic properties
+	if (test_bit (flags, KEY_CP_NAME))
 	{
-		dest->data.v = elektraStrNDup (source->data.v, source->dataSize);
-		if (!dest->data.v) goto memerror;
+		if (source->key)
+		{
+			dest->key = elektraStrNDup (source->key, source->keySize);
+			if (!dest->key) goto memerror;
+			dest->keySize = source->keySize;
+
+			ELEKTRA_ASSERT (source->ukey != NULL, "key != NULL but ukey == NULL");
+			dest->ukey = elektraStrNDup (source->ukey, source->keyUSize);
+			if (!dest->ukey) goto memerror;
+			dest->keyUSize = source->keyUSize;
+		}
+		else
+		{
+			dest->key = elektraStrDup ("/");
+			dest->keySize = 2;
+
+			dest->ukey = elektraMalloc (3);
+			dest->ukey[0] = KEY_NS_CASCADING;
+			dest->ukey[1] = '\0';
+			dest->ukey[2] = '\0';
+			dest->keyUSize = 3;
+		}
+		clear_bit (dest->flags, KEY_FLAG_MMAP_KEY);
 	}
-	else
+
+	if (test_bit (flags, KEY_CP_STRING))
 	{
-		dest->data.v = 0;
+		if (source->data.v)
+		{
+			dest->data.v = elektraStrNDup (source->data.v, source->dataSize);
+			if (!dest->data.v) goto memerror;
+			dest->dataSize = source->dataSize;
+
+			if (!test_bit (flags, KEY_META) && keyIsBinary (source))
+			{
+				keySetMeta (dest, "binary", "");
+			}
+		}
+		else
+		{
+			dest->data.v = NULL;
+			dest->dataSize = 0;
+		}
+		clear_bit (dest->flags, KEY_FLAG_MMAP_DATA);
 	}
 
-	if (source->meta)
+	if (test_bit (flags, KEY_CP_VALUE))
 	{
-		dest->meta = ksDup (source->meta);
-		if (!dest->meta) goto memerror;
+		if (source->data.v)
+		{
+			dest->data.v = elektraStrNDup (source->data.v, source->dataSize);
+			if (!dest->data.v) goto memerror;
+			dest->dataSize = source->dataSize;
+
+			if (!test_bit (flags, KEY_META) && keyIsBinary (source))
+			{
+				keySetMeta (dest, "binary", "");
+			}
+		}
+		else
+		{
+			dest->data.v = NULL;
+			dest->dataSize = 0;
+		}
+		clear_bit (dest->flags, KEY_FLAG_MMAP_DATA);
 	}
-	else
+
+	if (test_bit (flags, KEY_CP_META))
 	{
-		dest->meta = 0;
+		if (source->meta)
+		{
+			dest->meta = ksDup (source->meta);
+			if (!dest->meta) goto memerror;
+		}
+		else
+		{
+			dest->meta = 0;
+		}
 	}
 
 	// successful, now do the irreversible stuff: we obviously modified dest
 	set_bit (dest->flags, KEY_FLAG_SYNC);
 
-	// copy sizes accordingly
-	dest->keySize = source->keySize;
-	dest->keyUSize = source->keyUSize;
-	dest->dataSize = source->dataSize;
-
 	// free old resources of destination
-	if (!test_bit (dest->flags, KEY_FLAG_MMAP_KEY)) elektraFree (destKey);
-	if (!test_bit (dest->flags, KEY_FLAG_MMAP_KEY)) elektraFree (destUKey);
-	if (!test_bit (dest->flags, KEY_FLAG_MMAP_DATA)) elektraFree (destData);
-	ksDel (destMeta);
+	if (test_bit (flags, KEY_CP_NAME) && !test_bit (dest->flags, KEY_FLAG_MMAP_KEY)) elektraFree (orig.key);
+	if (test_bit (flags, KEY_CP_NAME) && !test_bit (dest->flags, KEY_FLAG_MMAP_KEY)) elektraFree (orig.ukey);
+	if (test_bit (flags, KEY_CP_VALUE) && !test_bit (dest->flags, KEY_FLAG_MMAP_DATA)) elektraFree (orig.data.c);
```

Comment:
I think this might have been a bug before you touched it. We obviously need to check the flags of the original key and **not** `dest->flags`, which may have been reset already. What surprises me is that this has been working.
Please correct me if I am wrong.
```suggestion
	if (test_bit (flags, KEY_CP_NAME) && !test_bit (orig.flags, KEY_FLAG_MMAP_KEY)) elektraFree (orig.key);
	if (test_bit (flags, KEY_CP_NAME) && !test_bit (orig.flags, KEY_FLAG_MMAP_KEY)) elektraFree (orig.ukey);
	if (test_bit (flags, KEY_CP_VALUE) && !test_bit (orig.flags, KEY_FLAG_MMAP_DATA)) elektraFree (orig.data.c);
```

Reply:
Yes, I think your right checking `orig` is safer. Checking `dest->flags` works too, because we don't call e.g. `keySetName` here, but modify e.g. `dest->key` directly, so we know that `dest->flags` wasn't changed.

Reply:
Actually the flags are literally reset before this check. Eg:
https://github.com/ElektraInitiative/libelektra/blob/3164e426da8f2c56d1879da0b90bd8b46f17b3ad/src/libs/elektra/key.c#L406

Reply:
Ah well, seems I added that :sweat_smile: 

On `master` the flags aren't cleared, so checking `dest->flags` wasn't a bug there.